Cwctl project list + project get --id/--name

james-wallis opened the pull request:
Implements https://github.com/eclipse/codewind/issues/1700
### Summary
* Add `cwctl project list` which returns all the projects on a Codewind instance (takes `--conid` to return information about remote instances)
* Add `cwctl project get --id/--name` to return information about a single project (`--conid` for remote again). User must supply either `--id` or `--name` (name will fetch the ID).
* Both of the new commands handle the `--json` global flag and also pretty print terminal output see below.
* Added error handling for 404 project not found on the `getProject` function, previously we outputted something about character `N` not able to be parsed.
* Ensure that an error is given back when either hostname or ports cannot be found for local Codewind. If an error is given when getting the hostname or ports, use that error.
* Refactored `GetProject` so that it can be mocked for testing.
```
❯ cwctl project ls

cwctl projecPROJECT ID 				NAME 		LANGUAGE 	APP STATUS 	LOCATION ON DISK
6e7e66b0-200e-11ea-9780-01852573fa4d	defaultNode	nodejs				/Users/james.wallis1@ibm.com/codewind-workspace/defaultNode
16ba0630-33a2-11ea-8fae-b389c50d3976	2020noddy	nodejs		Started		/Users/james.wallis1@ibm.com/codewind-workspace/2020noddy
5e613990-33a2-11ea-8fae-b389c50d3976	thursdaynoddy	nodejs		Started		/Users/james.wallis1@ibm.com/codewind-workspace/thursdaynoddy
```
```
❯ cwctl project get --id 6e7e66b0-200e-11ea-9780-01852573fa4d  

PROJECT ID 				NAME 		LANGUAGE 	APP STATUS 	LOCATION ON DISK
6e7e66b0-200e-11ea-9780-01852573fa4d	defaultNode	nodejs				/Users/james.wallis1@ibm.com/codewind-workspace/defaultNode

```
* Removed blank aliases from `Commands.go`.

### Testing
* Manual testing to ensure that both the commands for local and remote scenarios.
* Added Go test for getAll function.
* Added Go test for getProject and getProjectIDFromName.

liamchampton commented on pkg/actions/commands.go:
```diff
@@ -166,6 +166,31 @@ func Commands() {
+				},
+				{
+					Name:    "get",
+					Aliases: []string{},
```
If there is no aliases for the command this you dont need to have this line

james-wallis replied:
I've removed this and removed all other blank aliases.

markcor11 commented on pkg/project/getAll.go:
```diff
@@ -0,0 +1,62 @@
+		return nil, conErr.Err
+	}
+
+	client := &http.Client{}
```
Its ok to generate the client here, however I think we have been passing in the http client to the function so that we can call and mock it later if needed

markcor11 commented on pkg/project/getproject.go:
```diff
@@ -36,11 +42,15 @@ func GetProject(httpClient utils.HTTPClient, conID, projectID string) (*Project,
+		return nil, httpSecError
 	}
 	defer resp.Body.Close()
+	if resp.StatusCode == 404 {
```
rather than 404 you could also use the constants from the http package eg  http.StatusNotFound

james-wallis commented:
Since @markcor11 has reviewed this I have added functionality to do `cwctl project get --name` to make the command have a better user experience. In doing so I've changed the `unbind` command slightly by passing in the connection parameters.

markcor11 commented on pkg/actions/project.go:
```diff
@@ -135,3 +139,106 @@ func ProjectRemoveConnection(c *cli.Context) {
+
+	conURL, conErr := config.PFEOriginFromConnection(conInfo)
+	if conErr != nil {
+		fmt.Println(conErr.Err.Error())
```
Might be worth checking for JSON global flag and outputting utils.PrettyPrintJSON(conErr) to get the error in a consumable form or using the HandleConnectionError function to deal with formatting for you.

markcor11 commented on pkg/actions/project.go:
```diff
@@ -135,3 +139,106 @@ func ProjectRemoveConnection(c *cli.Context) {
+
+	conInfo, conInfoErr := connections.GetConnectionByID(conID)
+	if conInfoErr != nil {
+		fmt.Println(conInfoErr.Err.Error())
```
Might be worth checking for JSON global flag and outputting utils.PrettyPrintJSON(conErr) to get the error in a consumable form or using the HandleConnectionError function to deal with formatting for you.

markcor11 commented on pkg/actions/project.go:
```diff
@@ -135,3 +139,106 @@ func ProjectRemoveConnection(c *cli.Context) {
+
+	projects, err := project.GetAll(http.DefaultClient, conInfo, conURL)
+	if err != nil {
+		fmt.Println(err.Error())
```
Might be worth checking for JSON global flag and outputting utils.PrettyPrintJSON(conErr) you can't use HandleProjectError here because project.GetAll returns a go err.  Could project.GetAll return a projectErr ?

james-wallis commented:
> Could you change to use ProjectErr as mentioned inline and also handle JSON output please.

I've made this change @markcor11

jcockbain commented on pkg/actions/commands.go:
```diff
@@ -62,9 +62,8 @@ func Commands() {
-					Aliases: []string{""},
-					Usage:   "create a project on disk",
+					Name:  "create",
+					Usage: "create a project on disk",
```
Nit: we have mismatch of capitals on "usage" fields, more seem to be capital than not though. I'd make all these capitalised

james-wallis replied:
changed all for project

jcockbain commented on pkg/actions/project.go:
```diff
@@ -135,3 +139,106 @@ func ProjectRemoveConnection(c *cli.Context) {
+		HandleProjectError(getAllErr)
+		os.Exit(1)
+	}
+	if printAsJSON {
```
could have a blank line before if printAsJSON, for consistency?

james-wallis replied:
couple new lines added

jcockbain commented on pkg/actions/project.go:
```diff
@@ -135,3 +139,106 @@ func ProjectRemoveConnection(c *cli.Context) {
+	projectName := c.String("name")
+
+	if projectID == "" && projectName == "" {
+		fmt.Println("Error: Must specify either project ID (--id) or project name (--name)")
```
create a projectError here?

james-wallis replied:
have altered the way this works but not outputting as JSON as the CLI help does not either

jcockbain commented on pkg/actions/project.go:
```diff
@@ -135,3 +139,106 @@ func ProjectRemoveConnection(c *cli.Context) {
+	}
+
+	if projectID == "" && projectName != "" {
+		newProjectID, projectNameErr := project.GetProjectIDFromName(http.DefaultClient, conInfo, conURL, projectName)
```
Could have a project.getProjectFromName and a project.getProjectFromID

james-wallis replied:
Done this

jcockbain commented on pkg/actions/project.go:
```diff
@@ -135,3 +139,106 @@ func ProjectRemoveConnection(c *cli.Context) {
+		fmt.Println(string(json))
+	} else {
+		w := new(tabwriter.Writer)
+		w.Init(os.Stdout, 0, 8, 0, '\t', 0)
```
padding here

jcockbain commented on pkg/project/getAll.go:
```diff
@@ -0,0 +1,51 @@
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+
+package project
```
getproject.go v getAll.go is inconsistent

james-wallis replied:
changed to getall.go + test file

jcockbain commented on pkg/project/getproject.go:
```diff
@@ -16,41 +15,57 @@ type (
 	return &project, nil
 }
+
+// GetProjectIDFromName : Get a project ID using its name
```
projects?
